tests/ObjectReadStateMachineTests.cs

```diff
@@ -0,0 +1,308 @@
+using State = ObjectReadStateMachine.State;
+
+public sealed class ObjectReadStateMachineTests
+{
+    [Fact]
+    public void Default_Instance_Is_Initialized()
+    {
+        var subject = new ObjectReadStateMachine();
+        Assert.Equal(State.Initial, subject.CurrentState);
+    }
+
+    [Fact]
+    public void OnPropertyNameRead_Throws_When_CurrentState_Is_Not_PendingPropertyNameRead()
+    {
+        var subject = new ObjectReadStateMachine();
+        _ = Assert.Throws<InvalidOperationException>(subject.OnPropertyNameRead);
+    }
+
+    [Fact]
+    public void OnPropertyValueRead_Throws_When_CurrentState_Is_Not_PendingPropertyValueRead()
+    {
+        var subject = new ObjectReadStateMachine();
+        _ = Assert.Throws<InvalidOperationException>(subject.OnPropertyValueRead);
+    }
+
+    [Theory]
+    [InlineData(/*lang=json*/ "null")]
+    [InlineData(/*lang=json*/ "true")]
+    [InlineData(/*lang=json*/ "false")]
+    [InlineData(/*lang=json*/ "42")]
+    [InlineData(/*lang=json*/ """ "foobar" """)]
+    [InlineData(/*lang=json*/ "[]")]
+    public void Read_With_Invalid_Input_Is_An_Error(string json)
+    {
+        var subject = new ObjectReadStateMachine();
+        var reader = new Utf8JsonReader(Encoding.UTF8.GetBytes(json));
+
+        var result = subject.Read(ref reader);
+
+        Assert.Equal(ReadResult.Error, result);
+        Assert.Equal(State.Error, subject.CurrentState);
+    }
+
+    [Fact]
+    public void Read_Throws_When_State_Machine_Is_In_Error_State()
+    {
+        var subject = new ObjectReadStateMachine();
+        var reader = new Utf8JsonReader("["u8);
+
+        _ = subject.Read(ref reader);
+
+        Assert.Equal(State.Error, subject.CurrentState);
+
+        var ex = CatchExceptionForAssertion(reader, r => _ = subject.Read(ref r));
+        _ = Assert.IsType<InvalidOperationException>(ex);
+    }
+
+    [Fact]
+    public void Read_Throws_When_State_Machine_Is_In_Done_State()
+    {
+        var subject = new ObjectReadStateMachine();
+        var reader = new Utf8JsonReader("{}"u8);
+
+        _ = subject.Read(ref reader);
+
+        Assert.Equal(State.Done, subject.CurrentState);
+
+        var ex = CatchExceptionForAssertion(reader, r => _ = subject.Read(ref r));
+        _ = Assert.IsType<InvalidOperationException>(ex);
+    }
+
+    [Fact]
+    public void Read_Increments_CurrentPropertyLoopCount_When_Current_State_Is_PendingPropertyNameRead()
+    {
+        var subject = new ObjectReadStateMachine();
+        var reader = new Utf8JsonReader("{\"foo\":"u8, isFinalBlock: false, new());
+
+        foreach (var _ in Enumerable.Range(0, 10))
+        {
+            var result = subject.Read(ref reader);
+
+            Assert.Equal(ReadResult.PropertyName, result);
+            Assert.Equal(State.PendingPropertyNameRead, subject.CurrentState);
+        }
+    }
+
+    [Fact]
+    public void Read_Increments_CurrentPropertyLoopCount_When_Current_State_Is_PendingPropertyValueRead()
```

Comment:
`CurrentPropertyLoopCount` is no more?

Reply:
Removed with 89e05149e8d31ce6b73f781fd09a95d370f3cfea